command-line/lib/Command/Script.hs

```diff
@@ -0,0 +1,162 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE QuasiQuotes #-}
+
+{-# OPTIONS_HADDOCK hide #-}
+
+module Command.Script
+    ( Cmd
+    , mod
+    , run
+
+    -- * Internal
+    , requireSignatureOfParser
+    , requireAllOfParser
+    , requireAnyOfParser
+    , requireAtLeastOfParser
+    , scriptParser
+    ) where
+
+import Cardano.Multisig
+    ( Script (..), keyHashFromBytes, toScriptHash, validateScript )
+import Codec.Binary.Bech32.TH
+    ( humanReadablePart )
+import Codec.Binary.Encoding
+    ( fromBase16 )
+import Data.Char
+    ( isDigit, isLetter )
+import Data.Word
+    ( Word8 )
+import Options.Applicative
+    ( CommandFields, Mod, command, footerDoc, header, helper, info, progDesc )
+import Options.Applicative.Encoding
+    ( Encoding, encodingOpt )
+import Options.Applicative.Help.Pretty
+    ( bold, indent, string, vsep )
+import Prelude hiding
+    ( mod )
+import System.IO
+    ( stdin, stdout )
+import System.IO.Extra
+    ( hPutBytes, noNewline, progName )
+import Text.ParserCombinators.ReadP
+    ( ReadP, readP_to_S, (<++) )
+
+import qualified Data.ByteString.Char8 as B8
+import qualified Data.Text as T
+import qualified Data.Text.Encoding as T
+import qualified Text.ParserCombinators.ReadP as P
+
+
+newtype Cmd = Cmd
+    { encoding :: Encoding
+    } deriving (Show)
+
+mod :: (Cmd -> parent) -> Mod CommandFields parent
+mod liftCmd = command "scripthash" $
```

Comment:
Why not have a hash command nested under the  Script command group instead? `scripthash` as a command name sounds weird.

Reply:
so instead `cardano-address scripthash` have `cardano-address script` ?

Reply:
`cardano-address script hash` to be consistent with `cardano-address key hash`

Reply:
And also, leaves the room open for future commands on scripts.

Reply:
we will need to have also probably something like this for creating addreses:
1. (payment credential from script + stake credential from stake key) 
```
$ cat script.txt \
| cardano-address script hash
| cardano-address script payment --network-tag 1
| cardano-address script delegation $(cat stake.prv | cardano-address key public) 
```
2. (payment credential from script + stake credential from script) 
```
$ cat script1.txt \
| cardano-address script hash
| cardano-address script payment --network-tag 1
| cardano-address script delegation $(cat script2.txt | cardano-address script hash) 
```
3. (payment credential from key + stake credential from script) 
```
$ root.prv \
| cardano-address key child 1852H/1815H/0H/0/1 \
| cardano-address key public \
| cardano-address address payment --network-tag 1 \
| cardano-address script delegation $(cat script2.txt | cardano-address script hash) 
```

Reply:
@paweljakubas 

```
$ cardano-address script hash $(cat script.txt)
| cardano-address script payment --network-tag 1
| cardano-address script delegation $(cat stake.prv | cardano-address key public) 
```

```
$ cardano-address script hash $(cat script1.txt)
| cardano-address script payment --network-tag 1
| cardano-address script delegation $(cardano-address script hash $(cat script2.txt))
```

:grimacing:

Reply:
The benefits we get reading keys from stdin on other command is because there are many commands returning keys on stdout. For scripts, there's no command (nor do I foresee any use-case for it) returning a raw script as output.

Reply:
or we can also leave `cardano-address address` but we need a way to differentiate in Shelley impl between key credential and script credential...The former is hashed, the latter is hashed already. maybe we can detect this by inspcting the length of payload (bytestring)

Reply:
(sorry, I now realize I missed the point in your comment :facepalm: ...). I think  `script payment` or `script delegation` are rather confusing. We should use the `address` command group for creating addresses. But now, we need to make it possible to accept script hash or public keys.

Reply:
make "script hash" here https://github.com/input-output-hk/cardano-addresses/pull/76/commits/1db43edf071092430ff22d777390ffa6bd738f7d

Reply:
also when it comes how script hash is accommodated in payment/stake credential I planned to deal with it in forthcoming task  https://github.com/input-output-hk/cardano-addresses/pull/76 ending this U/S